### PR TITLE
customized api/v1/invoices query parameters to filter results

### PR DESCRIPTION
**Pull request by SakerOmera**

### CODE REVIEW

File: BTCPayServer/Controllers/GreenField/InvoiceController.cs [hunk 1]
```diff
@@ -47,19 +48,37 @@ public class GreenFieldInvoiceController : Controller
         [Authorize(Policy = Policies.CanViewInvoices,
             AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
         [HttpGet("~/api/v1/stores/{storeId}/invoices")]
-        public async Task<IActionResult> GetInvoices(string storeId, bool includeArchived = false)
+        public async Task<IActionResult> GetInvoicesV2(string storeId, string orderId, string status,
+            DateTimeOffset? startDate = null,
+            DateTimeOffset? endDate = null,
+            bool includeAddresses = true, bool includeArchived = false)
```
Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/InvoiceController.cs:54]: Please remove the `include Addresses` option as it doesn't fit this API entity

File: BTCPayServer/Controllers/GreenField/InvoiceController.cs [hunk 2]
```diff
@@ -47,19 +48,37 @@ public class GreenFieldInvoiceController : Controller
         [Authorize(Policy = Policies.CanViewInvoices,
             AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
         [HttpGet("~/api/v1/stores/{storeId}/invoices")]
-        public async Task<IActionResult> GetInvoices(string storeId, bool includeArchived = false)
+        public async Task<IActionResult> GetInvoicesV2(string storeId, string orderId, string status,
```
Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/InvoiceController.cs:51]: make orderId a string array

Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/InvoiceController.cs:51]: do not rename method

File: BTCPayServer/Controllers/GreenField/InvoiceController.cs [hunk 3]
```diff
@@ -47,19 +48,37 @@ public class GreenFieldInvoiceController : Controller
...
+            DateTimeOffset? startDate = null,
```
Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/InvoiceController.cs:52]: We have not yet decided how to handle dates in query strings. My gut tells me we need to handle them in the same format as we do in the json body, which is a unix timestamp in seconds. We may need to create some special attribute to tell the method parameter to be accepted in this way

**Review by Kukks — CHANGES_REQUESTED** [review]
Will also need to be documented in the swagger once we agree on the mentioned changes

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 4]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
...
+            DateTimeOffset? startDate = null,
+            DateTimeOffset? endDate = null,
+            bool includeAddresses = true, bool includeArchived = false,
             CancellationToken token = default)
         {
+            Dictionary<string, object> queryPayload = new Dictionary<string, object>();
+            queryPayload.Add(nameof(includeArchived), includeArchived);
+
+            if (startDate != null)
+                queryPayload.Add(nameof(startDate), startDate);
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:22]:
```suggestion
                queryPayload.Add(nameof(startDate), Utils.DateTimeToUnixTime(startDate));
```

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 5]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
...
+
+            if (endDate != null)
+                queryPayload.Add(nameof(endDate), endDate);
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:25]:
```suggestion
                queryPayload.Add(nameof(endDate), Utils.DateTimeToUnixTime(endDate));

```

**Comment by NicolasDorier** [comment]
the test `CheckSwaggerIsConformToSchema` is failing.

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 6]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
 {
     public partial class BTCPayServerClient
     {
-        public virtual async Task<IEnumerable<InvoiceData>> GetInvoices(string storeId, bool includeArchived = false,
+        public virtual async Task<IEnumerable<InvoiceData>> GetInvoices(string storeId, string orderId = null, string status = null,
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:12]: `status` should be an array of `InvoiceStatus`

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 7]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
...
+                queryPayload.Add(nameof(startDate), startDate);
+
+            if (endDate != null)
+                queryPayload.Add(nameof(endDate), endDate);
+
+            if (orderId != null)
+                queryPayload.Add(nameof(orderId), orderId);
+
+            if (status != null)
+                queryPayload.Add(nameof(status), status);
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:31]: Once `status` is an array of an enum as a parameter to the method, the provided value here should be converter to string[]

Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:31]:
```suggestion
                queryPayload.Add(nameof(status), status.Select(s=> s.ToString()).ToArray());
```

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 8]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
...
+
+            queryPayload.Add(nameof(includeAddresses), includeAddresses);
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:33]: remove

File: BTCPayServer.Client/BTCPayServerClient.Invoices.cs [hunk 9]
```diff
@@ -9,13 +9,33 @@ namespace BTCPayServer.Client
 {
     public partial class BTCPayServerClient
     {
-        public virtual async Task<IEnumerable<InvoiceData>> GetInvoices(string storeId, bool includeArchived = false,
+        public virtual async Task<IEnumerable<InvoiceData>> GetInvoices(string storeId, string orderId = null, string status = null,
+            DateTimeOffset? startDate = null,
+            DateTimeOffset? endDate = null,
+            bool includeAddresses = true, bool includeArchived = false,
```
Comment (Kukks) [inline BTCPayServer.Client/BTCPayServerClient.Invoices.cs:15]: remove `includeAddresses`

File: BTCPayServer/Controllers/GreenField/InvoiceController.cs [hunk 10]
```diff
@@ -47,7 +48,9 @@ public class GreenFieldInvoiceController : Controller
         [Authorize(Policy = Policies.CanViewInvoices,
             AuthenticationSchemes = AuthenticationSchemes.Greenfield)]
         [HttpGet("~/api/v1/stores/{storeId}/invoices")]
-        public async Task<IActionResult> GetInvoices(string storeId, bool includeArchived = false)
+        public async Task<IActionResult> GetInvoices(string storeId, [FromQuery] string[] orderId = null, [FromQuery] string[] status = null,
+            [FromQuery] DateTimeOffset? startDate = null,
```
Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/InvoiceController.cs:52]: startDate and endDate should be a `long?` and then converted in the method body (unless you introduce some Model Binder attribute to convert it behind the scenes)

File: BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json [hunk 11]
```diff
@@ -15,6 +15,51 @@
...
+                    },
+                    {
+                        "name": "status",
+                        "in": "query",
+                        "required": false,
+                        "description": "Array of statuses of invoices to be fetched",
+                        "schema": {
+                            "type": "array",
+                            "items": {
+                                "type": "string",
```
Comment (Kukks) [inline BTCPayServer/wwwroot/swagger/v1/swagger.template.invoices.json:39]: use the defined schema for the status instead:
`"$ref": "#/components/schemas/InvoiceStatus"`

**Review by Kukks — CHANGES_REQUESTED** [review]

**Comment by rockstardev** [comment]
Thanks for contribution @SakerOmera

**Comment by NicolasDorier** [comment]
test failing is unrelated

**Comment by NicolasDorier** [comment]
Pushed dcc4214dcb72aa6b8e2a5c51e36da8850181f69e to make things more easy to use

**Pull request merged**